File: sdk/go/pkg/neo4j/neo4j.go
/*
 * SPDX-FileCopyrightText: © Hypermode Inc. <hello@hypermode.com>
 * SPDX-License-Identifier: Apache-2.0
 */

package neo4j

import (
	"encoding/json"
	"fmt"
	"time"

	"github.com/hypermodeinc/modus/sdk/go/pkg/utils"
)

type Neo4jOption func(*neo4jOptions)

type neo4jOptions struct {
	dbName string
}

func WithDbName(dbName string) Neo4jOption {
	return func(o *neo4jOptions) {
		o.dbName = dbName
	}
}

// The result of a Neo4j query.
type EagerResult struct {
	Keys    []string
	Records []*Record
}

// A record in a Neo4j query result.
type Record struct {
	Keys   []string
	Values []string
}

// A type constraint for retrieving values from a Neo4j record.
type RecordValue interface {
	bool | int64 | float64 | string | time.Time |
		[]byte | []any | map[string]any |
		Node | Relationship | Path | Point2D | Point3D
}

type Entity interface {
	GetElementId() string
	GetProperties() map[string]any
}

type Node struct {
	ElementId string
	Labels    []string
	Props     map[string]any
}

func (n *Node) GetElementId() string {
	return n.ElementId
}

func (n *Node) GetProperties() map[string]any {
	return n.Props
}

// A relationship in a Neo4j query result.
type Relationship struct {
	ElementId      string
	StartElementId string
	EndElementId   string
	Type           string
	Props          map[string]any
}

func (r *Relationship) GetElementId() string {
	return r.ElementId
}

func (r *Relationship) GetProperties() map[string]any {
	return r.Props
}

type Path struct {
	Nodes         []Node
	Relationships []Relationship
}

// A type constraint for retrieving property values from a Neo4j entity.
type PropertyValue interface {
	bool | int64 | float64 | string | time.Time |
		[]byte | []any | Point2D | Point3D
}

// Point2D represents a two dimensional point in a particular coordinate reference system.
type Point2D struct {
	X float64
	Y float64

	// Id of coordinate reference system.
	SpatialRefId uint32
}

// Point3D represents a three dimensional point in a particular coordinate reference system.
type Point3D struct {
	X float64
	Y float64
	Z float64

	// Id of coordinate reference system.
	SpatialRefId uint32
}

// String returns string representation of this point.
func (p Point2D) String() string {
	return fmt.Sprintf("Point{SpatialRefId=%d, X=%f, Y=%f}", p.SpatialRefId, p.X, p.Y)
}

// String returns string representation of this point.
func (p Point3D) String() string {
	return fmt.Sprintf("Point{SpatialRefId=%d, X=%f, Y=%f, Z=%f}", p.SpatialRefId, p.X, p.Y, p.Z)
}

// Executes a query or mutation on the Neo4j database.
func ExecuteQuery(connection, query string, parameters map[string]any, opts ...Neo4jOption) (*EagerResult, error) {
	dbOpts := &neo4jOptions{
		dbName: "neo4j",
	}

	for _, opt := range opts {
		opt(dbOpts)
	}

	bytes, err := utils.JsonSerialize(parameters)
	if err != nil {
		return nil, err
	}

	parametersJson := string(bytes)

	response := hostExecuteQuery(&connection, &dbOpts.dbName, &query, &parametersJson)

	return response, nil
}

// Get a value from a record at a given key and cast or decode it to a specific type.
func GetRecordValue[T RecordValue](record *Record, key string) (T, error) {
	var val T
	for i, k := range record.Keys {
		if k == key {
			err := json.Unmarshal([]byte(record.Values[i]), &val)
			if err != nil {
				return *new(T), err
			} else {
				return val, nil
			}
		}
	}
	return *new(T), fmt.Errorf("key not found in record")

}

func (r *Record) Get(key string) (string, bool) {
	for i, k := range r.Keys {
		if k == key {
			return r.Values[i], true
		}
	}
	return "", false
}

func (r *Record) AsMap() map[string]string {
	result := make(map[string]string)
	for i, k := range r.Keys {
		result[k] = r.Values[i]
	}
	return result
}

func (r *Record) JSONMarshal() ([]byte, error) {
	result := "{"
	for i, k := range r.Keys {
		keyBytes, err := utils.JsonSerialize(k)
		if err != nil {
			return nil, err
		}
		result += fmt.Sprintf("%s:%s", keyBytes, r.Values[i])
		if i < len(r.Keys)-1 {
			result += ","
		}
	}
	result += "}"
	return []byte(result), nil
}

// Get a property from an entity at a given key and cast or decode it to a specific type.
func GetProperty[T PropertyValue](e Entity, key string) (T, error) {
	var val T
	rawVal, ok := e.GetProperties()[key]
	if !ok {
		return *new(T), fmt.Errorf("key not found in node")
	}
	switch any(val).(type) {
	case int64:
		float64Val, ok := rawVal.(float64)
		if !ok {
			return *new(T), fmt.Errorf("expected value to have type int64 but found type %T", rawVal)
		}
		return any(int64(float64Val)).(T), nil
	default:
		val, ok = rawVal.(T)
		if !ok {
			zeroValue := *new(T)
			return zeroValue, fmt.Errorf("expected value to have type %T but found type %T", zeroValue, rawVal)
		}
		return val, nil
	}

}

File: sdk/assemblyscript/src/assembly/neo4j.ts
/*
 * Copyright 2024 Hypermode Inc.
 * Licensed under the terms of the Apache License, Version 2.0
 * See the LICENSE file that accompanied this code for further details.
 *
 * SPDX-FileCopyrightText: 2024 Hypermode Inc. <hello@hypermode.com>
 * SPDX-License-Identifier: Apache-2.0
 */

import { JSON } from "json-as";
import { NamedParams as Variables } from "./database";
import { DynamicMap } from "./dynamicmap";
export { Variables };

// @ts-expect-error: decorator
@external("modus_neo4j_client", "executeQuery")
declare function hostExecuteQuery(
  connection: string,
  dbName: string,
  query: string,
  parametersJson: string,
): EagerResult;

/**
 * Executes a Cypher query on the Neo4j database.
 * @param connection - the name of the connection
 * @param query - the query to execute
 * @param parameters - the parameters to pass to the query
 * @param dbName - the name of the database
 * @returns The EagerResult from the Neo4j server
 */
export function executeQuery(
  connection: string,
  query: string,
  parameters: Variables = new Variables(),
  dbName: string = "neo4j",
): EagerResult {
  const paramsJson = parameters.toJSON();
  const response = hostExecuteQuery(connection, dbName, query, paramsJson);
  if (!response) {
    throw new Error("Error executing Query.");
  }

  return response;
}

/**
 * The result of a Neo4j query.
 */
@json
export class EagerResult {
  keys: string[] = [];
  records: Record[] = [];

  /**
   * @deprecated use `keys` (lowercase) instead
   */
  get Keys(): string[] {
    return this.keys;
  }

  /**
   * @deprecated use `records` (lowercase) instead
   */
  get Records(): Record[] {
    return this.records;
  }
}

/**
 * A record in a Neo4j query result.
 */
@json
export class Record {
  keys: string[] = [];
  values: string[] = [];

  /**
   * @deprecated use `keys` (lowercase) instead
   */
  get Keys(): string[] {
    return this.keys;
  }
  set Keys(value: string[]) {
    this.keys = value;
  }

  /**
   * @deprecated use `values` (lowercase) instead
   */
  get Values(): string[] {
    return this.values;
  }
  set Values(value: string[]) {
    this.values = value;
  }

  /**
  /* Get a value from a record at a given key as a JSON encoded string.
   */
  get(key: string): string {
    for (let i = 0; i < this.Keys.length; i++) {
      if (this.Keys[i] == key) {
        return this.Values[i];
      }
    }
    throw new Error("Key not found in record.");
  }

  /**
   * Get a value from a record at a given key and cast or decode it to a specific type.
   */
  getValue<T>(key: string): T {
    if (
      isInteger<T>() ||
      isFloat<T>() ||
      isBoolean<T>() ||
      isString<T>() ||
      idof<T>() === idof<Node>() ||
      idof<T>() === idof<Relationship>() ||
      idof<T>() === idof<Path>() ||
      idof<T>() === idof<DynamicMap>() ||
      idof<T>() === idof<string[]>() ||
      idof<T>() === idof<i8[]>() ||
      idof<T>() === idof<i16[]>() ||
      idof<T>() === idof<i32[]>() ||
      idof<T>() === idof<i64[]>() ||
      idof<T>() === idof<u8[]>() ||
      idof<T>() === idof<u16[]>() ||
      idof<T>() === idof<u32[]>() ||
      idof<T>() === idof<u64[]>() ||
      idof<T>() === idof<f32[]>() ||
      idof<T>() === idof<f64[]>() ||
      idof<T>() === idof<Date>() ||
      idof<T>() === idof<JSON.Raw>() ||
      idof<T>() === idof<JSON.Raw[]>() ||
      idof<T>() === idof<Point2D>() ||
      idof<T>() === idof<Point3D>()
    ) {
      for (let i = 0; i < this.keys.length; i++) {
        if (this.keys[i] == key) {
          return JSON.parse<T>(this.Values[i]);
        }
      }
      throw new Error("Key not found in record.");
    }
    throw new Error("Unsupported type.");
  }

  asMap(): Map<string, string> {
    const map = new Map<string, string>();
    for (let i = 0; i < this.keys.length; i++) {
      map.set(this.keys[i], this.values[i]);
    }
    return map;
  }


  @serializer
  serialize(self: Record): string {
    let out = "{";
    const end = self.keys.length - 1;
    for (let i = 0; i < end; i++) {
      const key = JSON.stringify(self.keys[i]);
      out += key + ":" + self.values[i] + ",";
    }

    if (end >= 0) {
      const key = JSON.stringify(self.keys[end]);
      out += key + ":" + self.values[end];
    }

    return out + "}";
  }


  @deserializer
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  deserialize(data: string): Record {
    throw new Error("Not implemented.");
  }
}


@json
abstract class Entity {

  @alias("ElementId")
  elementId!: string;


  @alias("Props")
  props!: DynamicMap;

  /**
   * @deprecated use `elementId` (lowercase) instead
   */
  get ElementId(): string {
    return this.elementId;
  }
  set ElementId(value: string) {
    this.elementId = value;
  }

  /**
   * @deprecated use `props` (lowercase) instead
   */
  get Props(): DynamicMap {
    return this.props;
  }
  set Props(value: DynamicMap) {
    this.props = value;
  }

  getProperty<T>(key: string): T {
    if (
      isInteger<T>() ||
      isFloat<T>() ||
      isBoolean<T>() ||
      isString<T>() ||
      idof<T>() === idof<Date>() ||
      idof<T>() === idof<JSON.Raw>() ||
      idof<T>() === idof<Point2D>() ||
      idof<T>() === idof<Point3D>()
    ) {
      return this.Props.get<T>(key);
    }
    throw new Error("Unsupported type.");
  }
}


@json
export class Node extends Entity {

  @alias("Labels")
  labels!: string[];

  /**
   * @deprecated use `labels` (lowercase) instead
   */
  get Labels(): string[] {
    return this.labels;
  }
  set Labels(value: string[]) {
    this.labels = value;
  }
}


@json
export class Relationship extends Entity {

  @alias("StartElementId")
  startElementId!: string;


  @alias("EndElementId")
  endElementId!: string;


  @alias("Type")
  type!: string;

  /**
   * @deprecated use `startElementId` (lowercase) instead
   */
  get StartElementId(): string {
    return this.startElementId;
  }
  set StartElementId(value: string) {
    this.startElementId = value;
  }

  /**
   * @deprecated use `endElementId` (lowercase) instead
   */
  get EndElementId(): string {
    return this.endElementId;
  }
  set EndElementId(value: string) {
    this.endElementId = value;
  }

  /**
   * @deprecated use `type` (lowercase) instead
   */
  get Type(): string {
    return this.type;
  }
  set Type(value: string) {
    this.type = value;
  }
}


@json
export class Path {

  @alias("Nodes")
  nodes!: Node[];


  @alias("Relationships")
  relationships!: Relationship[];

  /**
   * @deprecated use `nodes` (lowercase) instead
   */
  get Nodes(): Node[] {
    return this.nodes;
  }
  set Nodes(value: Node[]) {
    this.nodes = value;
  }

  /**
   * @deprecated use `relationships` (lowercase) instead
   */
  get Relationships(): Relationship[] {
    return this.relationships;
  }
  set Relationships(value: Relationship[]) {
    this.relationships = value;
  }
}


@json
export class Point2D {

  @alias("X")
  x!: f64;


  @alias("Y")
  y!: f64;


  @alias("SpatialRefId")
  spatialRefId!: u32;

  /**
   * @deprecated use `x` (lowercase) instead
   */
  get X(): f64 {
    return this.x;
  }
  set X(value: f64) {
    this.x = value;
  }

  /**
   * @deprecated use `y` (lowercase) instead
   */
  get Y(): f64 {
    return this.y;
  }
  set Y(value: f64) {
    this.y = value;
  }

  /**
   * @deprecated use `spatialRefId` (lowercase) instead
   */
  get SpatialRefId(): u32 {
    return this.spatialRefId;
  }
  set SpatialRefId(value: u32) {
    this.spatialRefId = value;
  }

  toString(): string {
    return `Point{SpatialRefId=${this.SpatialRefId}, X=${this.X}, Y=${this.Y}}`;
  }

  /**
   * @deprecated use `toString` instead
   */
  String(): string {
    return this.toString();
  }
}


@json
export class Point3D {

  @alias("X")
  x!: f64;


  @alias("Y")
  y!: f64;


  @alias("Z")
  z!: f64;


  @alias("SpatialRefId")
  spatialRefId!: u32;

  /**
   * @deprecated use `x` (lowercase) instead
   */
  get X(): f64 {
    return this.x;
  }
  set X(value: f64) {
    this.x = value;
  }

  /**
   * @deprecated use `y` (lowercase) instead
   */
  get Y(): f64 {
    return this.y;
  }
  set Y(value: f64) {
    this.y = value;
  }

  /**
   * @deprecated use `z` (lowercase) instead
   */
  get Z(): f64 {
    return this.z;
  }
  set Z(value: f64) {
    this.z = value;
  }

  /**
   * @deprecated use `spatialRefId` (lowercase) instead
   */
  get SpatialRefId(): u32 {
    return this.spatialRefId;
  }
  set SpatialRefId(value: u32) {
    this.spatialRefId = value;
  }

  toString(): string {
    return `Point{SpatialRefId=${this.SpatialRefId}, X=${this.X}, Y=${this.Y}, Z=${this.Z}}`;
  }

  /**
   * @deprecated use `toString` instead
   */
  String(): string {
    return this.toString();
  }
}
